File: tools/src/main/java/com/qeapp/tools/process/Process.java
package com.qeapp.tools.process;

import com.qeapp.tools.fraction.Fraction;

import org.jetbrains.annotations.NotNull;

/**
 * Abstract process
 */
public interface Process extends Fraction {

    interface Observer {
        default void onProcessStarted(Process process) {}
        default void onProcessProgress(Process process, double fraction) {}
        default void onProcessSuccess(Process process) {}
        default void onProcessCancelled(Process process) {}
    }
    interface Observable extends Process {
        void addProcessObserver(@NotNull Observer observer);
        void removeProcessObserver(Observer observer);
    }

    /**
     * Start process
     */
    void start();

    /**
     * Cancel process
     */
    void cancel();

    /**
     * Process is success
     */
    boolean isSuccess();
    /**
     * Process is active
     */
    boolean isActive();
    /**
     * Process is canceled
     */
    boolean isCanceled();

    /**
     * Current progress
     */
    @Override
    double getFraction();

}

File: tools/src/main/java/com/kekadoc/tools/process/Process.kt
package com.kekadoc.tools.process

import com.kekadoc.tools.fraction.Fraction
import com.kekadoc.tools.fraction.FractionValue

/**
 * Abstract process
 */
interface Process : Fraction {

    interface Observer {
        fun onProcessStarted(process: Process?) {}
        fun onProcessProgress(process: Process?, fraction: Double) {}
        fun onProcessSuccess(process: Process?) {}
        fun onProcessCancelled(process: Process?) {}
    }
    interface Observable : Process {
        fun addProcessObserver(observer: Observer)
        fun removeProcessObserver(observer: Observer?)
    }

    /**
     * Start process
     */
    fun start()

    /**
     * Cancel process
     */
    fun cancel()

    /**
     * Process is success
     */
    val isSuccess: Boolean

    /**
     * Process is active
     */
    val isActive: Boolean

    /**
     * Process is canceled
     */
    val isCanceled: Boolean

    /**
     * Current progress
     */
    override fun getFraction(): Double

}